[SYCL][ESIMD] Add new way to define main test and verification functions

vasilytric opened the pull request:
This feature lets our provide verification logic for unusual test cases from outside of `test` struct

In this PR was done the following updates:
- if some test uses reference data and `NumElems == 1`, then run test with each one value from reference data, but is cases when `NumElems > 1` the tests will be launched with whole reference data 
- `run_test` structure will be specific for each `test.cpp`-file

vasilytric committed:
[SYCL][ESIMD] Add new way to define main test and verification function

vasilytric committed:
[SYCL][ESIMD] Update comments for test on defaul simd constructor

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/common.hpp:
```diff
@@ -29,54 +29,50 @@ using shared_allocator = sycl::usm_allocator<DataT, sycl::usm::alloc::shared>;
+                        const std::string &data_type) {
+  bool passed{true};
+
+  for (size_t i = 0; i < result_data.size(); ++i) {
+    if (!are_bitwise_equal(ref_data[i], result_data[i])) {
+      passed = false;
+      log::fail<NumElems>(
+          "Simd by " + TestCaseT::get_description() +
+              " failed, retrieved: " + std::to_string(result_data[i]) +
+              ", expected: " + std::to_string(ref_data[i]) +
+              ", at index: " + std::to_string(i),
+          data_type);
+    }
```
Are all these changes about moving out the following lines?
```
  for (size_t i = 0; i < result_data.size(); ++i) {
    if (!are_bitwise_equal(ref_data[i], result_data[i])) {
```

The `log::fail` message will differ from test group to test group significantly. For example, conversion constructor would need to output both the source and destination types, the simd_mask or simd_view tests would have different template parameters - not to mention free functions. Or what if you have 2 vectors to compare for each case?

What do you think about moving out a `log::fail` message construction into a test description class instead? It may:

- have an abstract interface;
- result in some `log::fail(const ITestDescription&)` interface with no `NumElems` at all;
- provide a possibility to have generic `TestDescription: public ITestDescription` in a `ctors/common.hpp`
- moreover, you might even have `TestCaseT` either to be inherited from `TestDescription` or to have it encapsulated, if that's appropriate

vasilytric replied:
I've applied this comment in this commit e44cd0f653ddd3f6c73aa85230716eb71cad054e
I hope that I correctly understood you

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_default.cpp:
```diff
@@ -84,34 +82,49 @@ struct const_ref {
-      }
-    }
+
+    passed = verify_test_result<DataT, NumElems, TestCaseT>(result, data_type);
```
You are providing a custom `Verifier` and a custom `verify_test_result` in the same source file. What's the intent behind moving a part of this function into the separate one within the same source file?

vasilytric replied:
`verify_test_result` was removed and verification was moved into `Verifier` struct
removing was [here](https://github.com/intel/llvm-test-suite/pull/614/commits/e44cd0f653ddd3f6c73aa85230716eb71cad054e#diff-6d44d6893e30c56cec1c2e698ea43849081d8b092f7a87532bd567294d14ed69L85-L106) emplacing was [here](https://github.com/intel/llvm-test-suite/pull/614/commits/e44cd0f653ddd3f6c73aa85230716eb71cad054e#diff-6d44d6893e30c56cec1c2e698ea43849081d8b092f7a87532bd567294d14ed69R103-R110)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/common.hpp:
```diff
@@ -29,54 +29,50 @@ using shared_allocator = sycl::usm_allocator<DataT, sycl::usm::alloc::shared>;
 // Using functor class to be able to iterate over the pre-defined data types.
 template <typename DataT, int NumElems, typename TestCaseT> struct test {
-  bool operator()(sycl::queue &queue, const std::string &data_type) {
+  template <typename VerifierT>
```
So if you are using the template parameter anyway, why do we need to instantiate the instance?
Or why not just use an abstract base class?

vasilytric replied:
I've applied this comment [here](https://github.com/intel/llvm-test-suite/pull/614/commits/e44cd0f653ddd3f6c73aa85230716eb71cad054e#diff-a1a30a06ba7da50ce8e13a9af9caec10d385910429792a1de956a7f15cc3d646R68-R71)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/common.hpp:
```diff
@@ -29,54 +29,50 @@ using shared_allocator = sycl::usm_allocator<DataT, sycl::usm::alloc::shared>;
+    if constexpr (NumElems == 1) {
+      for (size_t i = 0; i < ref_data.size(); ++i) {
+        passed = verifier.template verify<DataT, NumElems, TestCaseT>(
+            queue, {ref_data[i]}, data_type);
```
Do we need to call `std::vector` constructor only because of `verify_test_result` signature here? Or there is anything I'm missing from the first glance.

vasilytric replied:
No, previously we didn't created `std::vector` with single element in case `NumElems == 1`

yuriykoch replied:
Looks like I didn't get an idea why do we need to split the `run_test` and `VerifierT` in a first place. They still looks tightly-coupled to me. Would you mind to provide an example of a case when we either change the logic for `run_test` without changing the logic for `VerifierT`, or we change the logic for `VerifierT` without changing the logic for `run_test`?

vasilytric replied:
The `VerifierT` will be differ for the each test: for test on copy constructor, the `simd` required other simd object, but for the fill constructor requires pre-defined base value and step value
For many simd constructors requires different tests results verification logic

But for all tests, except test on `simd` default constructor I expect that we will have the same `run_test` function

yuriykoch replied:
So for example, how will you get the different `base` and `step` values for the fill constructor out of the `{ref_data[i]}` provided for the `simd<T,1>` case?

I agree, different tests might need different verification logic. So let's summarize pros and cons so far:

- we might be able to not duplicate the following 6 lines in each `Verifier` (which superseeds the original `test` now):
```
std::vector<DataT> ref_data{generate_ref_data<DataT, NumElems>()};
if constexpr (NumElems == 1) {
  for (size_t i = 0; i < ref_data.size(); ++i) {
  }
} else {
}
```
- we are going to have additional level of abstraction, resulting in `for_all_types_and_dims<run_test, ..., Verifier>(types, dims, queue);` instead of `for_all_types_and_dims<run_test, ...>(types, dims, queue);` in very test
- we are going to have additional vector constructor call for every test case for `simd<T,1>`
- we are unable to change the `for-loop` with the `single_task` inside into the `parallel_for` in a future
- most importantly, everyone reading the test (on a failure investigation for example) should consider the logic of `run_test`, `Verifer` and the call stack to get idea of the test logic

Aren't we going to have much more restrictions after such change than we gain from it?

vasilytric replied:
The previous solution was deleted, each test will have own `run_test` function, i've done it in this commit: 565f89ada9619b6ce9fc0cf6358a82a6fe707a73

yuriykoch reviewed (commented):
Would  you mind to provide a description for reasons of such change needed?

Generally, I would suggest to use encapsulation of utils/helpers instead of hierarchical levels of abstractions to make code loosely coupled.

vasilytric committed:
Merge branch 'intel' into new_way_define_ver

vasilytric committed:
[SYCL][ESIMD] Update test architecture

- Verification algorithm state in "verify" function
- Add TestDescription class for constructing error message

vasilytric committed:
[SYCL][ESIMD] Use simd::copy_to() instead of returning simd

This lets avoid using copy or move constructor, if used constructor will
be broken, then all tests could failed

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/common.hpp:
```diff
@@ -29,54 +29,61 @@ using shared_allocator = sycl::usm_allocator<DataT, sycl::usm::alloc::shared>;
-  queue.submit([&](sycl::handler &cgh) {
-    const auto ref = ref_data.data();
-    auto out = result.data();
+template <typename DataT, int NumElems = 0>
```
```suggestion
template <typename DataT, int NumElems>
```
That's a helper class specific for the `simd<T,N>` constructors, so looks like we don't need such special case. Or do we?

```suggestion
template <typename DataT, int NumElems, typename ContextT>
```
Moreover, looks like we are going to specify different contexts for every constructor covered, so we may move the `ContextT::get_description()` call out as well. What do you think?

vasilytric replied:
This comment was applied [here](https://github.com/intel/llvm-test-suite/pull/614/commits/f5fcba9f39f6f63e6e4c8d2de3b4ea8cc64a0444#diff-a1a30a06ba7da50ce8e13a9af9caec10d385910429792a1de956a7f15cc3d646R32)

vasilytric committed:
[SYCL][ESIMD] Apply clang-format

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/common.hpp:
```diff
@@ -29,54 +29,61 @@ using shared_allocator = sycl::usm_allocator<DataT, sycl::usm::alloc::shared>;
-    const auto ref = ref_data.data();
-    auto out = result.data();
+template <typename DataT, int NumElems = 0>
+class TestDescription : public ITestDescription {
+protected:
+  const std::string m_data_type;
```
```suggestion
class TestDescription : public ITestDescription {
  const std::string m_data_type;
```

Are you going to have inheritance further more and to access these elements in child? Looks like it would result in duplication of the `to_string()` logic. I assume I could suggest something  - but we would need a specific example to discuss :)

vasilytric replied:
I don't know how, but I expect that `m_data_type` and other fields would be re-used in classes-inheritances
But they are `protected` and don't accessible by `.` or `->` operators

I don't have a specific example, all my examples - it's just imaginations :)
If it necessary I can make this fields are `private`

yuriykoch replied:
Let's start with the `private` for now

vasilytric replied:
I've done it in this commit: a46c51bc932553cab7ea460e9442943784a0eaf4

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/common.hpp:
```diff
@@ -29,54 +29,61 @@ using shared_allocator = sycl::usm_allocator<DataT, sycl::usm::alloc::shared>;
+  const std::string m_test_case_description;
+  DataT m_retrieved_val;
+  DataT m_expected_val;
+  size_t m_index;
```
```suggestion
  const DataT m_retrieved_val;
  const DataT m_expected_val;
  const size_t m_index;
```
Minor - we might make explicit that constructor handles all the initialization

vasilytric replied:
This comment was applied [here](https://github.com/intel/llvm-test-suite/pull/614/commits/f5fcba9f39f6f63e6e4c8d2de3b4ea8cc64a0444#diff-a1a30a06ba7da50ce8e13a9af9caec10d385910429792a1de956a7f15cc3d646R36-R38)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/common.hpp:
```diff
@@ -29,54 +29,61 @@ using shared_allocator = sycl::usm_allocator<DataT, sycl::usm::alloc::shared>;
+                  const std::string &data_type, DataT retrieved_val,
+                  DataT expected_val, size_t index)
+      : m_data_type(data_type), m_test_case_description(test_case_description),
+        m_retrieved_val(retrieved_val), m_expected_val(expected_val),
```
```suggestion
  TestDescription(const std::string &data_type, DataT retrieved_val,
                  DataT expected_val, size_t index)
      : m_data_type(data_type), m_retrieved_val(retrieved_val), 
        m_expected_val(expected_val),
```
We would need no `get_description()` call on the upper level if we add the `typename ContextT` template parameter to the struct itself. Do you have any concerns?

Also, just minor suggestion, feel free to ignore, but what would you say to have `TestDescription(i, retrieved, expected, type_name)` sequence for the call parameters? So we might have type name always on the last place (the same way it's required for `for_all_types`, so we might just make it consistent over the codebase).

vasilytric replied:
This comment was applied [here](https://github.com/intel/llvm-test-suite/pull/614/commits/f5fcba9f39f6f63e6e4c8d2de3b4ea8cc64a0444#diff-a1a30a06ba7da50ce8e13a9af9caec10d385910429792a1de956a7f15cc3d646R40-R45)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/common.hpp:
```diff
@@ -29,54 +29,61 @@ using shared_allocator = sycl::usm_allocator<DataT, sycl::usm::alloc::shared>;
+        m_index(index) {}
+
+  std::string to_string() const override {
+    // TODO: convert fp types into hex data type
```
```suggestion
    // TODO: Make strings for fp values more short during failure output, may be by using hex representation
```
What would you say?

vasilytric replied:
Good update. This comment was applied [here](https://github.com/intel/llvm-test-suite/pull/614/commits/f5fcba9f39f6f63e6e4c8d2de3b4ea8cc64a0444#diff-a1a30a06ba7da50ce8e13a9af9caec10d385910429792a1de956a7f15cc3d646R47-R49)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/common.hpp:
```diff
@@ -29,54 +29,61 @@ using shared_allocator = sycl::usm_allocator<DataT, sycl::usm::alloc::shared>;
+    if constexpr (NumElems != 0) {
+      log_msg += ", with num elements: ";
+      log_msg += std::to_string(NumElems);
+    }
```
(1) Looks like we don't need a separate case for zero NumElems, what do you think?
(2) Minor, feel free to ignore, but by using `operator+()` we might split string construction over multiple lines and have lesser git diffs in case of changes
(3) Also minor, but we might just print `simd<T,N>` with actual type and dimensions
(4) One more minor: you are using `test_case_description` and a `Simd by`. may be just state that's a context explicitly?

So, as a possible suggestion:
```suggestion
    std::string log_msg("Failed for simd<");

    log_msg += m_data_type + "," + std::to_string(NumElems) + ">";
    log_msg += ", with context: " + ContextT::get_description();
    log_msg += ", retrieved: " + std::to_string(m_retrieved_val);
    log_msg += ", expected: " + std::to_string(m_expected_val);
    log_msg += ", at index: " + std::to_string(m_index);
```

Moreover, I have a feeling we don't need an index at all.. It has a different meaning for the `simd<T,1>` tests. Wouldn't it be a bit strange to retrieve something like this:
`Failed for simd<int,1> ... at index 4`. What do you think?

vasilytric replied:
Perfect idea! :)
This comment was applied [here](https://github.com/intel/llvm-test-suite/pull/614/commits/f5fcba9f39f6f63e6e4c8d2de3b4ea8cc64a0444#diff-a1a30a06ba7da50ce8e13a9af9caec10d385910429792a1de956a7f15cc3d646L57-L61)

vasilytric committed:
[SYCL][ESIMD] Remove uniform initialization

To correspond of the project style

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_copy.cpp:
```diff
@@ -33,11 +33,11 @@ struct initializer {
 
   template <typename DataT, int NumElems>
-  static simd<DataT, NumElems> call_simd_ctor(const DataT *ref_data) {
+  static void call_simd_ctor(const DataT *ref_data, DataT *out) {
```
```suggestion
  static void call_simd_ctor(const DataT *const ref_data, DataT * const out) {
```
What would you say to make the constness of the pointer itself explicit? For each test case.

vasilytric replied:
I've applied this comment in this commit: https://github.com/intel/llvm-test-suite/pull/614/commits/e87ad2ec4b8cff0070719ae8992d50d244c2a019

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_copy.cpp:
```diff
@@ -77,35 +77,73 @@ class const_ref {
+// Struct that calls simd in provided context and then verifies obtained result.
+struct Verifier {
+  template <typename DataT, int NumElems, typename TestCaseT>
+  static bool verify(sycl::queue &queue, const std::vector<DataT> ref_data,
```
```suggestion
  static bool verify(sycl::queue &queue, const std::vector<DataT> &ref_data,
```
Did you mean this?

vasilytric replied:
I've applied this comment [here](https://github.com/intel/llvm-test-suite/commit/e87ad2ec4b8cff0070719ae8992d50d244c2a019#diff-6d44d6893e30c56cec1c2e698ea43849081d8b092f7a87532bd567294d14ed69L88-L89)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_copy.cpp:
```diff
@@ -77,35 +77,73 @@ class const_ref {
+
+    shared_vector<DataT> result(NumElems, shared_allocator<DataT>{queue});
+    shared_vector<DataT> shared_ref_data(ref_data.begin(), ref_data.end(),
+                                         shared_allocator<DataT>{queue});
```
```suggestion
    shared_allocator<DataT> allocator(queue);
    shared_vector<DataT> result(NumElems, allocator);
    shared_vector<DataT> reference(ref_data, allocator);
```
What about something like this?

vasilytric replied:
When I try construct the vector by this state: ` shared_vector<DataT> shared_ref_data(ref_data, allocator);` I got this error:

```
ctor_copy.cpp:105:26: error: no matching constructor for initialization of 'shared_vector<unsigned long long>' (aka 'vector<unsigned long long, usm_allocator<unsigned
      long long, sycl::usm::alloc::shared>>')
    shared_vector<DataT> shared_ref_data(ref_data, allocator);
                         ^               ~~~~~~~~~~~~~~~~~~~
```
To apply your comment `ref_data` should be `shared_vector` type, but it just `std::vector` type

I've applied this comment in this commit: https://github.com/intel/llvm-test-suite/pull/614/commits/e87ad2ec4b8cff0070719ae8992d50d244c2a019

yuriykoch replied:
Yep, sorry. It's a bit of surprise we cannot just state the source container with the only difference in allocator type.

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_copy.cpp:
```diff
@@ -77,35 +77,73 @@ class const_ref {
+
+    queue.submit([&](sycl::handler &cgh) {
+      const auto ref = shared_ref_data.data();
+      auto out = result.data();
```
```suggestion
      const DataT * const ref = shared_ref_data.data();
      const auto out = result.data();
```
The `const auto` gives the `DataT * const` type actually

vasilytric replied:
I've applied this comment in this commit: https://github.com/intel/llvm-test-suite/pull/614/commits/e87ad2ec4b8cff0070719ae8992d50d244c2a019

yuriykoch replied:
Looks like the `out` is still `auto` instead of the `const auto`, and I don't see any reasoning here. Is it intended or did it just slip by? Applies to `ctor_def` as well.

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_copy.cpp:
```diff
@@ -77,35 +77,73 @@ class const_ref {
+
+        auto test_description = TestDescription<DataT, NumElems>(
+            TestCaseT::get_description(), data_type, result[i], ref_data[i], i);
+        log::fail(test_description);
```
```suggestion
        const auto description = TestDescription<DataT, NumElems, TestCaseT>(
            i, result[i], ref_data[i], data_type);
        log::fail(description);
```
Minor: we might use a `const`; here and elsewhere.

vasilytric replied:
I've applied this comment [here](https://github.com/intel/llvm-test-suite/pull/614/commits/f5fcba9f39f6f63e6e4c8d2de3b4ea8cc64a0444#diff-e20a89556dfa9ec98f8eec4bea698de3975b780ff24a3bde1ced890af1d16581R122-R124)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_copy.cpp:
```diff
@@ -77,35 +77,73 @@ class const_ref {
+};
 
-  bool passed{true};
+int main(int argc, char **argv) {
```
```suggestion
int main(int, char **) {
```
To avoid warning about unused variables. Here and elsewhere.

vasilytric replied:
I've applied this comment [here](https://github.com/intel/llvm-test-suite/pull/614/commits/e87ad2ec4b8cff0070719ae8992d50d244c2a019#diff-e20a89556dfa9ec98f8eec4bea698de3975b780ff24a3bde1ced890af1d16581L132-L133)

yuriykoch replied:
Let's have it in a `ctors_default` as well

vasilytric replied:
oh, my mistake. was removed in this commit: e757cc8bfa62226cc21f28c8811647fed4c7cadb

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_copy.cpp:
```diff
@@ -77,35 +77,73 @@ class const_ref {
+  bool pass = true;
 
   const auto types{get_tested_types<tested_types::all>()};
   const auto dims{get_all_dimensions()};
```
```suggestion
  const auto types = get_tested_types<tested_types::all>();
  const auto dims = get_all_dimensions();
```
As a suggestion - it makes it more visually delimited, by virtually no additional cost.

vasilytric replied:
I've applied this comment [here](https://github.com/intel/llvm-test-suite/pull/614/commits/e87ad2ec4b8cff0070719ae8992d50d244c2a019#diff-e20a89556dfa9ec98f8eec4bea698de3975b780ff24a3bde1ced890af1d16581L138-L140)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_copy.cpp:
```diff
@@ -77,35 +77,73 @@ class const_ref {
-  passed &= for_all_types_and_dims<test, rval_in_expr>(types, dims, queue);
-  passed &= for_all_types_and_dims<test, const_ref>(types, dims, queue);
+  pass &=
+      for_all_types_and_dims<run_test, initializer, Verifier>(types, dims, q);
```
Is it possible to have some type alias to avoid `Verifier` here?

vasilytric replied:
I think no, I can't imagine some different alias for this struct

yuriykoch replied:
I've meant something like
```
template <typename DataT, int NumElems, typename TestCaseT>
using run_copy_test = run_test<DataT, NumElems, testCaseT, Verifier>;
```

It would make code dependency more explicit, still not sure it would solve the original issue

vasilytric replied:
in current implementation this alias is useless due to the fact that we have statements like this `passed &= for_all_types_and_dims<run_test, initializer>(types, dims, queue);`
it length 76 symbols

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_copy.cpp:
```diff
@@ -77,35 +77,73 @@ class const_ref {
-  std::cout << (passed ? "=== Test passed\n" : "=== Test FAILED\n");
-  return passed ? 0 : 1;
+  std::cout << (pass ? "=== Test passed\n" : "=== Test FAILED\n");
+  return pass ? 0 : 1;
```
```suggestion
  return ok ? 0 : 1;
```
Just as a suggestion, if you are looking for a shorter naming to use over the tests. Anyway, just be consistent with it - that's the same flag on the all levels anyway. So if you change it somewhere - let's change it everywhere.
Actually, the `passed` looks good to me, as it's hard to misread it as a verb instead of noun/adj.

vasilytric replied:
I make a better solution, I've done it in this commit: https://github.com/intel/llvm-test-suite/pull/614/commits/f67072db7233c3435d7b090999e75f79c333019b

yuriykoch replied:
OK, so we stick with the `passed` then :)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_default.cpp:
```diff
@@ -20,22 +20,20 @@
 using namespace cl::sycl;
 using namespace sycl::ext::intel::experimental::esimd;
+using namespace esimd_test::api::functional::ctors;
 using namespace esimd_test::api::functional;
```
```suggestion
```
Same for the `ctor_copy.cpp`

vasilytric replied:
When I apply this comment I got the following compilation error:
```
ctor_default.cpp:122:39: error: use of undeclared identifier 'tested_types'
  const auto types = get_tested_types<tested_types::all>();
ctor_default.cpp:123:21: error: use of undeclared identifier 'get_all_dimensions'; did you mean 'esimd_test::api::functional::get_all_dimensions'?
  const auto dims = get_all_dimensions();
                    ^~~~~~~~~~~~~~~~~~
                    esimd_test::api::functional::get_all_dimensions
SYCL\ESIMD\api\functional\ctors/../type_coverage.hpp:96:6: note: 'esimd_test::api::functional::get_all_dimensions' declared here
auto get_all_dimensions() { return values_pack<1, 8, 16, 32>(); }
```

yuriykoch replied:
Right, sorry. Looks like you've removed `using namespace esimd_test::api::functional::ctors;` instead. That's great, as it makes the `ctors::` namespace usage explicit :)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_default.cpp:
```diff
@@ -20,22 +20,20 @@
-#include "../common.hpp"
+#include "common.hpp"
 
 using namespace cl::sycl;
```
```suggestion
```
Not sure it's a good idea. What if we have name clash in `esimd_test` and `sycl`? Anyway, let's use the `sycl` namespace to align with SYCL-2020, not the `cl::sycl` one

vasilytric replied:
I've applied this comment in this commit: https://github.com/intel/llvm-test-suite/pull/614/commits/98e64557d3ac8bb9d222967a32688856bb78cc2d

yuriykoch replied:
Please, make it explicit if the suggestion doesn't fit well. So you've changed it to the `using namespace sycl;` - any ideas on what if we have name clash in esimd_test and sycl?

vasilytric replied:
this statement was deleted in this commit: ce4735d0c5c845a2eb948b1c59e84651f724961e

yuriykoch commented on SYCL/ESIMD/api/functional/logger.hpp:
```diff
@@ -20,22 +20,19 @@
 namespace esimd_test {
 namespace api {
 namespace functional {
```
```suggestion
namespace esimd_test::api::functional {
```
While we are here :)

vasilytric replied:
I've applied this comment in this commit: https://github.com/intel/llvm-test-suite/pull/614/commits/d22d5e27f2e4c39195c318acbf85851734ad9ef9

yuriykoch replied:
Great, thanks for the applying to all files as well :)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_default.cpp:
```diff
@@ -84,52 +82,53 @@ struct const_ref {
+// Struct that calls simd in provided context and then verifies obtained result.
+struct Verifier {
+  template <typename DataT, int NumElems, typename TestCaseT>
+  static bool verify(sycl::queue &queue, const std::vector<DataT> ref_data,
```
```suggestion
  static bool verify(sycl::queue &queue, const std::vector<DataT>,
```
We are not using it, right? One more reason to have the `Verifier` and `run_test` as the single instance.

Moreover, how many times will this function run with, for example, `simd<int,1>` and `initializer`? Looks like it would loop over several times in the `run_test`, wouldn't it?

vasilytric replied:
The `Verifier` in test on simd default constructor was replaced into `run_test` in this commit: https://github.com/intel/llvm-test-suite/pull/614/commits/9fa6675d9d2aa91003819b23c03915291ea76af2

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_default.cpp:
```diff
@@ -84,52 +82,53 @@ struct const_ref {
-    using AllocatorT = sycl::usm_allocator<DataT, sycl::usm::alloc::shared>;
 
-    std::vector<DataT, AllocatorT> result{NumElems, AllocatorT{queue}};
+    shared_vector<DataT> result(NumElems, shared_allocator<DataT>{queue});
```
```suggestion
    shared_vector<DataT> result(NumElems, shared_allocator<DataT>(queue));
```
Here and elsewhere

vasilytric replied:
I've applied this comment in this commit: e87ad2ec4b8cff0070719ae8992d50d244c2a019

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_default.cpp:
```diff
@@ -84,52 +82,53 @@ struct const_ref {
-      if (it != default_val) {
+
+    for (size_t i = 0; i < result.size(); ++i) {
+      if (result[i] != default_val) {
```
Once again, do we need to output index? The only concern I have is for the `simd<T,1>` failures output though.

vasilytric replied:
I think that all is ok, for example output in case `NumElements == 1`:
```
Failed for initializer failed, retrieved: 0, expected: 0, at index: 0, with data type: char, with num elements: 1
Failed for initializer failed, retrieved: 0, expected: 0, at index: 0, with data type: unsigned char, with num elements: 1
Failed for initializer failed, retrieved: 0, expected: 0, at index: 0, with data type: signed char, with num elements: 1
Failed for initializer failed, retrieved: 0, expected: 0, at index: 0, with data type: short, with num elements: 1
Failed for initializer failed, retrieved: 0, expected: 0, at index: 0, with data type: unsigned short, with num elements: 1
Failed for initializer failed, retrieved: 0, expected: 0, at index: 0, with data type: int, with num elements: 1
Failed for initializer failed, retrieved: 0, expected: 0, at index: 0, with data type: unsigned int, with num elements: 1
```

yuriykoch replied:
It's all right then) Sorry for disturbance.

vasilytric committed:
[SYCL][ESIMD] Update TestDescription

- New failure messages type is better, e.g.: "Failed for simd<int, 8>,
  with context: ..."
- Was changed expected obtained parameters

vasilytric committed:
[SYCL][ESIMD] Update tests for simd default and move constructors

- Save const correctness for pointers on data, which uses on device
- Shared allocator init as varable to not construct it twice
- "main(int argc, char **argv)" was replaced with "main(int, char **)"
  to avoid unexpected warnings

vasilytric committed:
[SYCL][ESIMD] Remove using run_test from ctors/common in test on default ctor

vasilytric committed:
[SYCL][ESIMD] Replace defining namespaces in few str with defining in one

vasilytric committed:
[SYCL][ESIMD] Use "using sycl" instead of "using cl::sycl"

vasilytric committed:
[SYCL][ESIMD] Replace variable names "queue" to "q" and "pass" to "passed"

To have the same formatting in main function

vasilytric commented:
> Looks like I didn't get an idea why do we need to split the run_test and VerifierT in a first place. They still looks tightly-coupled to me. 

@yuriykoch Yes, you is right they are tightly-coupled, but I do not want to copy-paste the following code in each test-file:
```
    // If current number of elements is equal to one, then run test with each
    // one value from reference data.
    // If current number of elements is greater than one, then run tests with
    // whole reference data.
    if constexpr (NumElems == 1) {
      for (size_t i = 0; i < ref_data.size(); ++i) {
        *** init necessary variables ***
        *** run test actions ***
        *** verify test result values ***
      }
    } else {
        *** init necessary variables ***
        *** run test actions ***
        *** verify test result values ***
    }
```
> Would you mind to provide an example of a case when we either change the logic for run_test without changing the logic for VerifierT, or we change the logic for VerifierT without changing the logic for run_test?

For the test on simd fill constructor we will have the same `run_test` logic, but we will have the differ `VerifierT` logic

vasilytric committed:
[SYCL][ESIMD] Rename missed variables

vasilytric committed:
[SYCL][ESIMD] Replace "auto" with "DataT *const" for output shared data

This change let's use define const correctness explicitly

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_copy.cpp:
```diff
@@ -77,34 +77,76 @@ class const_ref {
+
+    shared_allocator<DataT> allocator(queue);
+    shared_vector<DataT> result(NumElems, allocator);
+    shared_vector<DataT> shared_ref_data(ref_data.begin(), ref_data.end(),
+                                         allocator);
```
Looks like we are missing assertion on a `ref_data.size()` equality to `NumElems`

vasilytric replied:
assertion was added in this commit: a6fac3bb1b02498714bd0edbdd9648053e8a27f7

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_default.cpp:
```diff
@@ -78,57 +75,57 @@ struct const_ref {
-    using AllocatorT = sycl::usm_allocator<DataT, sycl::usm::alloc::shared>;
 
-    std::vector<DataT, AllocatorT> result{NumElems, AllocatorT{queue}};
+    ctors::shared_vector<DataT> result(NumElems,
```
Minor: looks like both the `shared_vector` and `shared_allocator` could be safely moved out to the `functional` from the `functional/ctors`

vasilytric replied:
I've applied this comment in this commit: cad22ffd112a0adfa68a2ac6aaa7a012d6b9c88c

yuriykoch commented on SYCL/ESIMD/api/functional/logger.hpp:
```diff
@@ -17,32 +17,26 @@
-  }
-  std::cout << log_msg << std::endl;
+inline void fail(const ITestDescription &test_description) {
+  std::cout << test_description.to_string() << std::endl;
```
```suggestion
  // Force output buffer flush after each failure
  std::cout << test_description.to_string() << std::endl;
```

vasilytric replied:
comment was added in this commit: d40351a9b5151861e6a8a0afd342413140ce1c2f

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/common.hpp:
```diff
@@ -29,59 +26,58 @@ using shared_allocator = sycl::usm_allocator<DataT, sycl::usm::alloc::shared>;
+    log_msg += ", with context: " + ContextT::get_description();
+    log_msg += ", retrieved: " + std::to_string(m_retrieved_val);
+    log_msg += ", expected: " + std::to_string(m_expected_val);
+    log_msg += ", at index: " + std::to_string(m_index);
```
One more minor suggestion. What if we print something like `Failed for simd<int,8>[1]` instead of `Failed for simd<int,8> ..., at index: 1`? Would it be more obscure, what do you think?

vasilytric replied:
`..., at index: 1` this style will using in one of the test on esimd

yuriykoch reviewed (commented):
> Yes, you is right they are tightly-coupled, but I do not want to copy-paste the following code in each test-file

I would suggest to have a `run_test` on per-test basis for now, with `Verifier` converted into the private member function `run_test::run_check` instead of a separate entities.

> For the test on simd fill constructor we will have the same run_test logic, but we will have the differ VerifierT logic

Looks like the `run_test` will not provide different `base` and `step` results for `simd<T,1>` case

Anyway, looks like this PR will provide a great fix for tests logic for a reference data usage for the `simd<T,1>` case

vasilytric commented:
> So for example, how will you get the different base and step values for the fill constructor out of the {ref_data[i]} provided for the simd<T,1> case?

@yuriykoch  For the fill constructor we will have the same case as for the default constructor

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/common.hpp:
```diff
@@ -29,59 +26,58 @@ using shared_allocator = sycl::usm_allocator<DataT, sycl::usm::alloc::shared>;
-                ", expected: " + std::to_string(ref_data[it]),
-            data_type);
+    bool passed = true;
+    std::vector<DataT> ref_data{generate_ref_data<DataT, NumElems>()};
```
```suggestion
    const std::vector<DataT> ref_data = generate_ref_data<DataT, NumElems>();
```

vasilytric replied:
I've applied this comment in this commit: a6fac3bb1b02498714bd0edbdd9648053e8a27f7

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_copy.cpp:
```diff
@@ -47,11 +47,11 @@ struct var_decl {
+  static void call_simd_ctor(const DataT *const ref_data, DataT *const out) {
     simd<DataT, NumElems> source_simd;
     source_simd.copy_from(ref_data);
     simd<DataT, NumElems> simd_by_var_decl{source_simd};
```
```suggestion
    simd<DataT, NumElems> simd_by_var_decl(source_simd);
```
We should ensure we are calling exactly the constructor required.

vasilytric replied:
I've applied this comment in this commit: a6fac3bb1b02498714bd0edbdd9648053e8a27f7

vasilytric committed:
[SYCL][ESIMD] Update test architecture

- run_test now defined in the test.cpp-file - its removes one
  abstraction level
- added assert that reference data size is equal to number of elements
  in simd
- {} was replaced with ()

vasilytric committed:
[SYCL][ESIMD] Move shared vector and allocator into functional/common.hpp

They may use in the future

vasilytric committed:
[SYCL][ESIMD] Remove "using namespace sycl" due to possible  names conflicts

vasilytric committed:
[SYCL][ESIMD] Remove argument names in main function

Due to may be unexpected warnings about unused variables

vasilytric committed:
[SYCL][ESIMD] Make fields of TestDescription class private

vasilytric committed:
[SYCL][ESIMD] Replace variable name "q" to "queue" in ctor_default.cpp

vasilytric committed:
[SYCL][ESIMD] Add comment in log:fail()

Explicitly say that output buffer will be forced flushed

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_default.cpp:
```diff
@@ -78,57 +74,56 @@ struct const_ref {
-  sycl::queue queue{esimd_test::ESIMDSelector{},
-                    esimd_test::createExceptionHandler()};
+int main(int, char **) {
+  sycl::queue queue(esimd_test::ESIMDSelector(),
```
```suggestion
  sycl::queue queue(esimd_test::ESIMDSelector{},
```
Minor, actually. Looks like all other tests use such way to initialize the ESIMDSelector instance, probably to visually differentiate the instantiation from the function call. What do you think? Here and in ctor_copy.

vasilytric replied:
I've done it in this commit: 06d104d6498d91c00d7acd83e697c6c9fd5c0856

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_copy.cpp:
```diff
@@ -23,7 +23,6 @@
-using namespace sycl;
 using namespace sycl::ext::intel::experimental::esimd;
 using namespace esimd_test::api::functional::ctors;
 using namespace esimd_test::api::functional;
```
```suggestion
using namespace esimd_test::api::functional;
```
Let's be consistent with the `ctors_default.cpp`

vasilytric replied:
I've done it in this commit: 77dd9fb4a5fdb0e2058cd42506413458ce4ec498

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_copy.cpp:
```diff
@@ -77,34 +76,95 @@ class const_ref {
+    // one value from reference data.
+    // If current number of elements is greater than one, then run tests with
+    // whole reference data.
+    if constexpr (NumElems == 1) {
```
I would  suggest to state the logic fix explicitly in PR description

vasilytric replied:
done

yuriykoch reviewed (commented):
Thanks for your patience and assistance, Vasily

vasilytric committed:
[SYCL][ESIMD] Remove using functional::ctors from ctor_copy.cpp

Do same as it implemented in test on default ctor

vasilytric committed:
[SYCL][ESIMD] Replace ESIMDSelector() with ESIMDSelector{}

vasilytric commented:
Looks like the latest pre-ci-windows failed due to the issue https://github.com/intel/llvm/issues/8816

@sndmitriev Would you mind to take a look on this PR?

vasilytric committed:
Merge branch 'intel' into new_way_define_ver

vasilytric commented:
linux CI failed due to the test `DiscardEvents/invalid_event.cpp` failed. This test is not related to the current PR

sndmitriev reviewed (approved):
Looks good to me.

vasilytric commented:
> Looks good to me.

@sndmitriev Thanks for the review. Could you merge this PR?

sndmitriev commented:
> > Looks good to me.
> 
> @sndmitriev Thanks for the review. Could you merge this PR?

Unfortunately I am not authorized to merge PRs with failing pre-commit checks. This has to be done by someone who is authorized to do it.

vasilytric commented:
> Unfortunately I am not authorized to merge PRs with failing pre-commit checks. This has to be done by someone who is authorized to do it.

@sndmitriev All checks have passed now. I hope that now you can merge this PR.

sndmitriev commented:
> > Unfortunately I am not authorized to merge PRs with failing pre-commit checks. This has to be done by someone who is authorized to do it.
> 
> @sndmitriev All checks have passed now. I hope that now you can merge this PR.

I do not know why, but I still see that I am not authorized to merge this PR. @vladimirlaz can you please help?
